### PR TITLE
FAQ for type inference

### PR DESCRIPTION
**Pull request by som-snytt**
Documents https://github.com/scala/bug/issues/7743

The tour still needs tweaking. per https://github.com/scala/docs.scala-lang/issues/2648

### CODE REVIEW

**Comment by SethTisue** [comment]
@som-snytt wdyt about the following version?

### Why couldn't Scala infer the correct type for my code?

This is difficult to generalize about, as there could be a variety of causes and possible solutions.

Type inference was greatly improved in Scala 3. If you're using Scala 2, you might try your example in Scala 3 to see if it's handled any better.

The most common workaround is simply to add an explicit type to your code.

Sometimes breaking up a parameter list into multiple sections helps, as per [this language tour section](https://docs.scala-lang.org/tour/multiple-parameter-lists.html#drive-type-inference).

Questions about type inference often involve `toSet`. For discussion of examples involving `toSet`, see:

* https://github.com/scala/bug/issues/7743
* https://stackoverflow.com/questions/5544536/in-scala-2-type-inference-fails-on-set-made-with-toset

**Comment by SethTisue** [comment]
@BalmungSan this might interest you as well

**Comment by som-snytt** [comment]
### If it's called dotty, then why doesn't my dotted expression work?

### If I'm implying the correct type for my code, then why can't Scala infer it?

  Are you using "imply" and "infer" correctly?

### Expectation or Anticipation

I see my PR was more of a placeholder.

The github states are draft (WIP, drafts cannot be merged) and open green (ready for review), so if you're reviewing, it should be green. I don't know why there isn't a third state for "shared draft" that precedes green. That would mean, even if somebody accidentally grants an accepting review, don't merge it, as we're still mulling.

**Comment by BalmungSan** [comment]
Personally, I really like this FAQ addition.

My proposal for the title may be:

## Why couldn't Scala infer the correct type for my dot-chain expression?

Additionally, I think providing a concrete example of the suggested solutions would be useful, as well as explaining why that fixes the problem.

Like:

```scala
// Chane this:
List("ab").toSet.head.head

// Into this:
val set = List("ab").toSet
set.head.head
// This works because ...

// Another option is:
List("ab").toSet[String].head.head
// This also works because ...
```

**Comment by SethTisue** [comment]
> Additionally, I think providing a concrete example of the suggested solutions would be useful, as well as explaining why that fixes the problem

I'm consciously choosing not to do this. In general, I seek to keep FAQ entries short and sweet. A brief answer, and then perhaps a links to whatever other sources on the Internet (or on our own documentation site) that expound further.

**Comment by BalmungSan** [comment]
> I'm consciously choosing not to do this. In general, I seek to keep FAQ entries short and sweet. A brief answer, and then a link or two or three to whatever other sources on the Internet

Oh right, forgot about that.
In that case I still expect whatever link is chosen to have such examples and explanations :)
Be it the official **Scala** documentation or a **StackOverflow** question / answer as other entries.

**Comment by som-snytt** [comment]
OK I will update with something like Seth's text. I'm having trouble with any title that says "dot". We say "dotted name" or similar for `com.acme.foo`, but "what do you call it when an expression uses a selection?" "Selection" seems too "specialized" a word for a FAQ. One reason this is a FAQ is that it's not obvious why I can't rewrite expressions in obvious ways and have it just work. You can say Liskov till you're blue in the face, referential transparency blah-blah, but if simple syntax tweaks break my code, then I will be confused and angry. I don't think a link to a ticket suffices, but a link to the Big Book of Scala can be supplied later.

**Comment by som-snytt** [comment]
ok I updated the section title.

### toSet or not toSet, that is the question

File: _overviews/FAQ/index.md [hunk 1]
```diff
@@ -196,18 +196,21 @@ be written using the `_` syntax.
...
-For more context, see [this section of the tour](https://docs.scala-lang.org/tour/multiple-parameter-lists.html#drive-type-inference).
+The most straightforward workaround is to supply explicit types in your code.
+
+That may involve specifying an explicit type to a definition, or a type argument to a method.
+
+Type inference is greatly improved in Scala 3. If Scala 2 doesn't compile your code, it's worth trying with Scala 3.
+
+Sometimes, using multiple parameter lists helps inference, as explained in [this section of the language tour](https://docs.scala-lang.org/tour/multiple-parameter-lists.html#drive-type-inference).
+
+For paradigmatic questions about type inference involving `toSet`, see the discussions on [this ticket](https://github.com/scala/bug/issues/7743) and a related [Q&A](https://stackoverflow.com/questions/5544536/in-scala-2-type-inference-fails-on-set-made-with-toset).
```
Comment (SethTisue) [inline _overviews/FAQ/index.md:213]: overall LGTM but I ask for one change: s/paradigmatic/common/

Comment (som-snytt) [reply _overviews/FAQ/index.md:213]: ok I accept pitching it to the commons

**Comment by som-snytt** [comment]
I neglected to mention that I inlined the links, to make it ASAP (short as preferred).

I changed `paradigmatic question` to `stupid` before the final edit.

**Review by BalmungSan — APPROVED** [review]

**Comment by som-snytt** [comment]
It should say, "If you are here, you should ask BalmungSan".

**Pull request merged**